File: src/test/java/com/meghneelgore/geometry/primitives/TestSegment.java
/*
 * Copyright (c) 2018.  Meghneel Gore (meghneel.gore@gmail.com)
 */

package com.meghneelgore.geometry.primitives;

import org.junit.Assert;
import org.junit.Test;

public class TestSegment {

    @Test(expected = IllegalArgumentException.class)
    public void testSegmentWithEqualPoints() {
        Segment s = new Segment(new Point(0, 1), new Point(0, 1));
    }

    @Test
    public void testSegment() {
        Point p1 = new Point(1.1, 2.2);
        Point p2 = new Point(2.2, 1.1);
        Segment s = new Segment(p1, p2);
        Assert.assertEquals("Initialized point different from retrieved", p1, s.getP1());
        Assert.assertEquals("Initialized point different from retrieved", p2, s.getP2());
    }

    @Test
    public void testLength() {
        Point p1 = new Point(1, 1);
        Point p2 = new Point(2, 2);
        Segment segment = new Segment(p1, p2);
        double length = segment.length();
        Assert.assertEquals("Calculated length is not right", 1.4142135623730951, length, 0.0);
    }

    @Test
    public void testIsParallelTo() {

        //Test slope branch
        Point p1 = new Point(0, 1);
        Point p2 = new Point(1, 0);
        Segment segment1 = new Segment(p1, p2);
        p1 = new Point(0, 2);
        p2 = new Point(2, 0);
        Segment segment2 = new Segment(p1, p2);

        Assert.assertTrue("Parallel segments calculated as non-parallel", segment1.isParallelTo(segment2));

        //Test vertical branch
        p1 = new Point(10, 1);
        p2 = new Point(10, 0);
        segment1 = new Segment(p1, p2);
        p1 = new Point(0, 2);
        p2 = new Point(0, 3);
        segment2 = new Segment(p1, p2);

        Assert.assertTrue("Parallel segments calculated as non-parallel", segment1.isParallelTo(segment2));

        //Test horizontal branch
        p1 = new Point(10, 1);
        p2 = new Point(100, 1);
        segment1 = new Segment(p1, p2);
        p1 = new Point(1, 2);
        p2 = new Point(-29, 2);
        segment2 = new Segment(p1, p2);

        Assert.assertTrue("Parallel segments calculated as non-parallel", segment1.isParallelTo(segment2));
    }

    @Test
    public void testIsPerpendicular() {
        //Test slope branch
        Point p1 = new Point(0, 1);
        Point p2 = new Point(1, 0);
        Segment segment1 = new Segment(p1, p2);
        p1 = new Point(0, 0);
        p2 = new Point(1, 1);
        Segment segment2 = new Segment(p1, p2);

        Assert.assertTrue("Perpendicular segments calculated as non-perpendicular", segment1.isPerpendicularTo(segment2));

        p1 = new Point(0, 1);
        p2 = new Point(1, 0);
        segment1 = new Segment(p1, p2);
        p1 = new Point(0, 0);
        p2 = new Point(1, 2);
        segment2 = new Segment(p1, p2);

        Assert.assertFalse("Non-perpendicular segments calculated as perpendicular", segment1.isPerpendicularTo(segment2));

    }

    @Test
    public void testIntersectsWith() {
        Point p1 = new Point(0, 1);
        Point p2 = new Point(1, 0);
        Segment segment1 = new Segment(p1, p2);
        p1 = new Point(0, 0);
        p2 = new Point(1, 1);
        Segment segment2 = new Segment(p1, p2);

        Assert.assertTrue("Intersecting segments calculated as non-intersecting", segment1.intersectsWith(segment2));

        p1 = new Point(0, 1);
        p2 = new Point(1, 0);
        segment1 = new Segment(p1, p2);
        p1 = new Point(10, 10);
        p2 = new Point(111, 111);
        segment2 = new Segment(p1, p2);

        Assert.assertFalse("Non-intersecting segments calculated as intersecting", segment1.intersectsWith(segment2));

        p1 = new Point(0, 0);
        p2 = new Point(3, 3);
        segment1 = new Segment(p1, p2);
        p1 = new Point(2, 2);
        p2 = new Point(4, 4);
        segment2 = new Segment(p1, p2);

        Assert.assertTrue("Intersecting segments calculated as non-intersecting", segment1.intersectsWith(segment2));

        p1 = new Point(1, 1);
        p2 = new Point(3, 1);
        segment1 = new Segment(p1, p2);
        p1 = new Point(2, 1);
        p2 = new Point(4, 7);
        segment2 = new Segment(p1, p2);

        Assert.assertTrue("Intersecting segments calculated as non-intersecting", segment1.intersectsWith(segment2));
    }

    @Test(expected = IllegalStateException.class)
    public void testSlopeForVerticalSegment() {
        Point p1 = new Point(1, 1);
        Point p2 = new Point(1, 2);
        Segment s1 = new Segment(p1, p2);
        s1.slope();
    }

    @Test
    public void testSlopeForNonVerticalSegment() {
        Point p1 = new Point(1, 1);
        Point p2 = new Point(2, 2);
        Segment s1 = new Segment(p1, p2);
        double slope = s1.slope();
        Assert.assertEquals("Calculated slope is wrong", 1.0, slope, 0.0);
    }

    @Test
    public void testFindAngleWith() {
        Point p1 = new Point(1, 1);
        Point p2 = new Point(2, 2);
        Segment s1 = new Segment(p1, p2);
        p1 = new Point(3, 3);
        p2 = new Point(4, 4);
        Segment s2 = new Segment(p1, p2);

        Assert.assertEquals("Angle calculated wrongly", 0.0, s1.findAngleWith(s2), 0.0);

        p1 = new Point(0, 0);
        p2 = new Point(2, 2);
        s1 = new Segment(p1, p2);
        p1 = new Point(2, 0);
        p2 = new Point(0, 2);
        s2 = new Segment(p1, p2);

        Assert.assertEquals("Angle calculated wrongly", Math.PI / 2, s1.findAngleWith(s2), 0.0);
    }

    @Test
    public void testEquals() {
        Point p1 = new Point(1, 1);
        Point p2 = new Point(2, 2);
        Segment s1 = new Segment(p1, p2);
        p1 = new Point(1, 1);
        p2 = new Point(2, 2);
        Segment s2 = new Segment(p1, p2);

        Assert.assertTrue("Equal segments not computed as equal", s1.equals(s2));
    }



}

File: src/test/kotlin/geometry/primitives/TestSegment.kt
/*
 * Copyright (c) 2018.  Meghneel Gore (meghneel.gore@gmail.com)
 */

package geometry.primitives

import geometry.primitives.Point
import geometry.primitives.Segment
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class TestSegment {

    @Test(expected = IllegalArgumentException::class)
    fun testSegmentWithEqualPoints() {
        val s = Segment(Point(0.0, 1.0), Point(0.0, 1.0))
    }

    @Test
    fun testSegment() {
        val p1 = Point(1.1, 2.2)
        val p2 = Point(2.2, 1.1)
        val s = Segment(p1, p2)
        assertEquals("Initialized point different from retrieved", p1, s.p1)
        assertEquals("Initialized point different from retrieved", p2, s.p2)
    }

    @Test
    fun testLength() {
        val p1 = Point(1.0, 1.0)
        val p2 = Point(2.0, 2.0)
        val segment = Segment(p1, p2)
        val length = segment.length
        assertEquals("Calculated getLength is not right", 1.4142135623730951, length, 0.0)
    }

    @Test
    fun testIsParallelTo() {
        //Test slope branch
        var p1 = Point(0.0, 1.0)
        var p2 = Point(1.0, 0.0)
        var segment1 = Segment(p1, p2)
        p1 = Point(0.0, 2.0)
        p2 = Point(2.0, 0.0)
        var segment2 = Segment(p1, p2)

        assertTrue("Parallel segments calculated as non-parallel", segment1.isParallelTo(segment2))

        //Test vertical branch
        p1 = Point(10.0, 1.0)
        p2 = Point(10.0, 0.0)
        segment1 = Segment(p1, p2)
        p1 = Point(0.0, 2.0)
        p2 = Point(0.0, 3.0)
        segment2 = Segment(p1, p2)

        assertTrue("Parallel segments calculated as non-parallel", segment1.isParallelTo(segment2))

        //Test horizontal branch
        p1 = Point(10.0, 1.0)
        p2 = Point(100.0, 1.0)
        segment1 = Segment(p1, p2)
        p1 = Point(1.0, 2.0)
        p2 = Point(-29.0, 2.0)
        segment2 = Segment(p1, p2)

        assertTrue("Parallel segments calculated as non-parallel", segment1.isParallelTo(segment2))
    }

    @Test
    fun testIsPerpendicular() {
        //Test slope branch
        var p1 = Point(0.0, 1.0)
        var p2 = Point(1.0, 0.0)
        var segment1 = Segment(p1, p2)
        p1 = Point(0.0, 0.0)
        p2 = Point(1.0, 1.0)
        var segment2 = Segment(p1, p2)

        assertTrue("Perpendicular segments calculated as non-perpendicular", segment1.isPerpendicularTo(segment2))

        p1 = Point(0.0, 1.0)
        p2 = Point(1.0, 0.0)
        segment1 = Segment(p1, p2)
        p1 = Point(0.0, 0.0)
        p2 = Point(1.0, 2.0)
        segment2 = Segment(p1, p2)

        Assert.assertFalse("Non-perpendicular segments calculated as perpendicular", segment1.isPerpendicularTo(segment2))

    }

    @Test
    fun testIntersectsWith() {
        var p1 = Point(0.0, 1.0)
        var p2 = Point(1.0, 0.0)
        var segment1 = Segment(p1, p2)
        p1 = Point(0.0, 0.0)
        p2 = Point(1.0, 1.0)
        var segment2 = Segment(p1, p2)

        assertTrue("Intersecting segments calculated as non-intersecting", segment1.intersectsWith(segment2))

        p1 = Point(0.0, 1.0)
        p2 = Point(1.0, 0.0)
        segment1 = Segment(p1, p2)
        p1 = Point(10.0, 10.0)
        p2 = Point(111.0, 111.0)
        segment2 = Segment(p1, p2)

        Assert.assertFalse("Non-intersecting segments calculated as intersecting", segment1.intersectsWith(segment2))

        p1 = Point(0.0, 0.0)
        p2 = Point(3.0, 3.0)
        segment1 = Segment(p1, p2)
        p1 = Point(2.0, 2.0)
        p2 = Point(4.0, 4.0)
        segment2 = Segment(p1, p2)

        assertTrue("Intersecting segments calculated as non-intersecting", segment1.intersectsWith(segment2))

        p1 = Point(1.0, 1.0)
        p2 = Point(3.0, 1.0)
        segment1 = Segment(p1, p2)
        p1 = Point(2.0, 1.0)
        p2 = Point(4.0, 7.0)
        segment2 = Segment(p1, p2)

        assertTrue("Intersecting segments calculated as non-intersecting", segment1.intersectsWith(segment2))
    }

    @Test
    fun testSlopeForVerticalSegment() {
        val p1 = Point(1.0, 1.0)
        val p2 = Point(1.0, 2.0)
        val s1 = Segment(p1, p2)
        assertEquals("Vertical segment slope calculated incorrectly", Double.POSITIVE_INFINITY, s1.slope, 0.0)
    }

    @Test
    fun testSlopeForNonVerticalSegment() {
        val p1 = Point(1.0, 1.0)
        val p2 = Point(2.0, 2.0)
        val s1 = Segment(p1, p2)
        val slope = s1.slope
        assertEquals("Calculated slope is wrong", 1.0, slope, 0.0)
    }

    @Test
    fun testFindAngleWith() {
        var p1 = Point(1.0, 1.0)
        var p2 = Point(2.0, 2.0)
        var s1 = Segment(p1, p2)
        p1 = Point(3.0, 3.0)
        p2 = Point(4.0, 4.0)
        var s2 = Segment(p1, p2)

        assertEquals("Angle calculated wrongly", 0.0, s1.angleWith(s2), 0.0)

        p1 = Point(0.0, 0.0)
        p2 = Point(2.0, 2.0)
        s1 = Segment(p1, p2)
        p1 = Point(2.0, 0.0)
        p2 = Point(0.0, 2.0)
        s2 = Segment(p1, p2)

        assertEquals("Angle calculated wrongly", -Math.PI / 2, s1.angleWith(s2), 0.0)
    }

    @Test
    fun testEquals() {
        var p1 = Point(1.0, 1.0)
        var p2 = Point(2.0, 2.0)
        val s1 = Segment(p1, p2)
        p1 = Point(1.0, 1.0)
        p2 = Point(2.0, 2.0)
        val s2 = Segment(p1, p2)

        assertTrue("Equal segments not computed as equal", s1 == s2)
    }


}
